### ISSUE TITLE
Invalid SAM CLI executable configured

### ISSUE THREAD

**Issue by dougfultz** [aws/aws-toolkit-jetbrains#577]
**Describe the bug**
SAM is detected by the plugin, but attempting to create a new `AWS Serverless Application` shows:
<img width="780" alt="screen shot 2018-11-21 at 1 45 22 pm" src="https://user-images.githubusercontent.com/14137949/48862821-71cd7380-ed95-11e8-952e-71019cc8bf56.png">

If I go into the settings and specify the sam path it shows more of the error.
<img width="1358" alt="screen shot 2018-11-21 at 1 46 02 pm" src="https://user-images.githubusercontent.com/14137949/48862842-827de980-ed95-11e8-92e6-8293120dd1c3.png">

**To reproduce**
1. Install the plugin
2. Attempt to create a new project
3. Manually set the SAM location in preferences

**Expected behavior**
I would expect a new project to be created or the settings to be saved.

**Your Environment**
 - OS: macOS 10.12.6
 - JetBrains' Product: PyCharm CE
 - JetBrains' Product Version: 2018.3
 - Toolkit Version: v0.3

**Comment by sanathkr** [comment]
@dougfultz What is the default language on your system? 

Also, can you paste output of `locale`?

**Comment by dougfultz** [comment]
English is the default language.
<img width="664" alt="screen shot 2018-11-21 at 6 47 08 pm" src="https://user-images.githubusercontent.com/14137949/48873820-f16f3880-edbd-11e8-9247-403b95804ec9.png">

```
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

**Comment by sanathkr** [comment]
That looks right. May be it has something to do with how the IDE is spawing the process. 

@kiiadi Can you run `locale` from the IDE toolkit and print the output from there?

**Comment by rli** [comment]
I have been able to replicate by doing the following:
1. Launch toolkit development IDE via terminal with the following: `bash -c 'LANG=POSIX LC_ALL=POSIX LC_CTYPE=POSIX open /Applications/IntelliJ\ IDEA.app/'`
1. Launch test PyCharm instance via the standard `:runIde` Gradle task
1. PyCharm instance now has an environment where it thinks the locale is POSIX, and passes this along to the SAM CLI.

This issue occurs due to some Mac-specific hacks.

The IDE tries to setup a sane environment by passing the parent environment on non-Mac environments. On Macs, if the IDE was launched via GUI (Finder, Dock, Spotlight etc.), the IDE's parent environment is practically useless, so the IDE tries to build an environment by running a Python script that dumps the environment (seriously. see [here](https://github.com/JetBrains/intellij-community/blob/master/bin/mac/printenv.py) and [here](https://github.com/JetBrains/intellij-community/blob/36f6e9d591452411382a63186820b107a5751170/platform/util/src/com/intellij/util/EnvironmentUtil.java#L181-L204).) If no environmental variable in the set {LANG, LC_ALL, LC_CTYPE} can be found, the IDE sets LC_CTYPE to `<location>.<JVM 'file.encoding' parameter>` with a fallback to "en_US.UTF-8". This behavior is highly likely to give us a UTF-8-compatible environment, (unless your terminal environment doesn't have the locale properly set, which is not the case here), and is the most likely case for users.

However, if you launch the IDE through a command-line mechanism, the above does not apply and inherits the environment from whatever launched it.

tl;dr: Looks like there are some potential issues if you don't launch PyCharm via a Mac built-in GUI (Finder, Dock, Spotlight etc.). @dougfultz, can you describe how you started PyCharm?

**Comment by sanathkr** [comment]
@rli great deep dive! Thanks for that!

**Comment by kiiadi** [comment]
@rli what's the behaviour if we launch from JetBrains Toolbox.

What's the proposed fix here? Should we be inspecting all the "parent" environment variables and removing the ones that'll make SAM croak?

**Comment by rli** [comment]
Dove into behavior a bit more, seems like the environment patch is default behavior on Mac, unless the user explicitly disables it via the IDE properties file.

I am not convinced that this is a problem with how the IDE invokes SAM, since I can tell the IDE to give the SAM CLI no environmental variables (`.withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.NONE)`), and the validation works just fine. (`env -i /usr/local/bin/sam` also works for me, and is the same idea: run SAM without any env vars)

**Comment by sanathkr** [comment]
But if you do that you are stripping out essential env like PATH which is needed by SAM build command to find the right executable to build using. 

I will discourage against no env vars mode. Instead it’s okay to manually set LANG=en_US.UTF8 since SAM CLI does not yet do localization

**Comment by rli** [comment]
In the short-term, I have no problem with forcing LANG=en_US.UTF-8, but I would like to root cause this before moving forward with a solution.

Another datapoint: I have managed to create an environment where `locale` is POSIX in the context of the IDE, but SAM validates fine....

**Comment by dougfultz** [comment]
@sanathkr I am launching PyCharm using the GUI launcher.

I can also confirm that launching PyCharm with the terminal resolves both errors I originally posted.

```
$ /Applications/PyCharm\ CE.app/Contents/MacOS/pycharm 
2018-11-26 10:03:15.819 pycharm[69399:1086327] allVms required 1.8*,1.8+
2018-11-26 10:03:15.823 pycharm[69399:1086338] Value of PYCHARM_VM_OPTIONS is (null)
2018-11-26 10:03:15.823 pycharm[69399:1086338] fullFileName is: /Applications/PyCharm CE.app/Contents/bin/pycharm.vmoptions
2018-11-26 10:03:15.823 pycharm[69399:1086338] fullFileName exists: /Applications/PyCharm CE.app/Contents/bin/pycharm.vmoptions
2018-11-26 10:03:15.823 pycharm[69399:1086338] Processing VMOptions file at /Applications/PyCharm CE.app/Contents/bin/pycharm.vmoptions
2018-11-26 10:03:15.824 pycharm[69399:1086338] Done
```

**Comment by rli** [comment]
@dougfultz We've made some headway on this.

Can you run `env | grep -E "(LANG|LC_.*?)=.*" -` and confirm that your output matches?
```
$ env | grep -E "(LANG|LC_.*?)=.*" -
LC_CTYPE=
LANG=en_US.UTF-8
```

**Comment by dougfultz** [comment]
```
$ env | grep -E "(LANG|LC_.*?)=.*" -
LANG=en_US.UTF-8
```

**Comment by cbilgin** [comment]
I'm observing the same error here too. 

locale on terminal gives 
`bilginb:~
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=`

locale on the IDE gives
`$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
`
Using this setting I get the following error 
<img width="1007" alt="screen shot 2018-11-30 at 7 03 51 pm" src="https://user-images.githubusercontent.com/22490017/49323598-badaa180-f4d2-11e8-8fb9-95217c582b57.png">


The proposed solution `bash -c 'LANG=POSIX LC_ALL=POSIX LC_CTYPE=POSIX open /Applications/IntelliJ\ IDEA.app/'` gives the following error
<img width="1006" alt="screen shot 2018-11-30 at 7 06 55 pm" src="https://user-images.githubusercontent.com/22490017/49323622-245ab000-f4d3-11e8-961e-480f62e5c26b.png">

**Comment by rli** [comment]
@dougfultz hmm, seems like you have a different problem then. The problem that we've observed is that oh-my-zsh does not export LANG by default, but sets LC_CTYPE based on LANG, resulting in `LC_CTYPE=`, which upsets the SAM CLI.

@cbilgin `bash -c 'LANG=POSIX LC_ALL=POSIX LC_CTYPE=POSIX open /Applications/IntelliJ\ IDEA.app/'` was not a proposed solution, but rather a way to trigger the same error seen in this issue. Your issue is that your SAM CLI does not have the `--info` command, which was introduced in SAM CLI version 0.7.0. Can you try updating to version 0.8.1 (currently the latest version as of Nov. 30, 2018)

**Comment by cbilgin** [comment]
I see, pip update is stuck in 0.6.2, I used brew install which gave the latest version. Thank you @rli

**Comment by josjaf** [comment]
I'm getting the same issue here too

**Comment by abrooksv** [comment]
Related YouTrack issue: https://youtrack.jetbrains.com/issue/PY-30780
